### PR TITLE
Experiments

### PR DESCRIPTION
**Pull request by rodolfocarobene**
Sorry for the chaos

Checklist before review:
- [ ] Test on hardware.
- [ ] Fix signal generator write/query.

Checklist before merge:
- [ ] Reviewers confirm new code works as expected.
- [ ] Tests are passing.
- [ ] Documentation is updated.

### CODE REVIEW

**Comment by rodolfocarobene** [comment]
@PietroCampana should we review this?

**Comment by PietroCampana** [comment]
It's almost done, but I still would like to add some data saving/loading functions and do more hardware tests

**Comment by PietroCampana** [comment]
> @PietroCampana should we review this?
> 

It's almost done, but I still would like to add some data saving/loading functions and do more hardware tests

File: src/qtics/experiment.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+                    futures.append(executor.submit(m.watch, self.event))
+                futures.append(executor.submit(self.main))
+                done, _ = wait(futures, return_when=FIRST_COMPLETED)
+                if len(done) > 0 and len(done) != len(futures):
+                    future = done.pop()
+                    if future.exception() is not None:
+                        log.warning(
+                            "One task failed with: %s, shutting down.",
+                            future.exception(),
+                        )
```
Comment (rodolfocarobene) [inline src/qtics/experiment.py:191]: In the case of exceptions the return is automatic? 
If the return is usable, I would also consider it for another de-nesting

File: tests/test_experiment.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    """Test run with monitor failure."""
+    monitor.max_read = 0
+    experiment.instrument1.update_defaults(name="reset occurred")
+    experiment.add_monitor(monitor)
+    experiment.run()
+    assert experiment.instrument1.name == "reset occurred"
+
+
+if __name__ == "__main__":
+    pytest.main()
```
Comment (rodolfocarobene) [inline tests/test_experiment.py:185]:
```suggestion
```

File: src/qtics/experiment.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+    def watch(self, event: Event):
+        """Run the experiment continuously until event is set."""
+        self.all_instruments("connect")
+        log.info("Running monitor %s", self.name)
+        while True:
+            time.sleep(self.sleep)
+            self.main()
+            if event.is_set():
+                log.info("Trigger event set, %s shutting down.", self.name)
+                return
```
Comment (rodolfocarobene) [inline src/qtics/experiment.py:156]:
```suggestion
        while True and not event.is_set():
            time.sleep(self.sleep)
            self.main()
        else:
            log.info("Trigger event set, %s shutting down.", self.name)
```
This removes a line, so it's an incredible improvement

File: src/qtics/experiment.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+            if parent_name != "":
+                parent_group = file.require_group(parent_name)
+                group = parent_group.require_group(group_name)
+            else:
+                group = file.require_group(group_name)
+
+            if datasets is not None:
+                for data_name, data in datasets.items():
+                    group.create_dataset(data_name, data=data)
+            if attributes:
```
Comment (rodolfocarobene) [inline src/qtics/experiment.py:103]: In general, it is recommended to not rely on "bare" truthfulness and falseness of arguments to check if they exist or not.

**Review by rodolfocarobene — COMMENTED** [review]
Thanks @PietroCampana for all the work! 
In general, the PR seems good. I have added some minor comments :-)

**Review by rodolfocarobene — COMMENTED** [review]
LGTM, consider this approved

**Review by PietroCampana — APPROVED** [review]

**Pull request merged**